### PR TITLE
support custom CA bundle for AWS API

### PR DESCRIPTION
**Pull request by staebler**
When a cluster is installed in a AWS C2S region, access to the AWS API requires a custom CA bundle for trust. The custom CA bundle is read from the "ca-bundle.pem" key of the kube-cloud-config ConfigMap in the openshift-config-managed namespace.

To use the custom CA bundle in the CSI driver pod, the operator will sync the kube-cloud-config ConfigMap from the openshift-config-managed namespace to the operator's namespace.

The CSI driver uses the CA bundle by having the AWS_CA_BUNDLE environment variable point to the CA bundle file. Unfortunately, the AWS SDK requires that, if the AWS_CA_BUNDLE environemnt variable is set, then the file must exist. We cannot simply always set the environment variable and rely on the presence or absence of the ConfigMap to determine the behavior. Consequently, the CSI driver deployment manifest has been templatized to fill out the parts needed to support a custom CA bundle only when there is a custom CA bundle.

https://issues.redhat.com/browse/CORS-1584

### CODE REVIEW

**Comment by staebler** [comment]
See https://github.com/openshift/cluster-storage-operator/pull/109 for the changes in the cluster-storage-operator to support this.

File: assets/controller.yaml [hunk 1]
```diff
@@ -52,6 +52,10 @@ spec:
               value: '1'
             - name: AWS_CONFIG_FILE
               value: /var/run/secrets/aws/credentials
+            {{- if .CABundleConfigMap}}
+            - name: AWS_CA_BUNDLE
+              value: /ca/ca-bundle.pem
```
[the remark this answers is not in the document]
Comment (staebler) [reply assets/controller.yaml:57]: Changed it to `/etc/ca/ca-bundle.pem`? Is that good?

**Comment by staebler** [comment]
I have re-worked this to move the ConfigMap syncing from cluster-storage-operator into here.

**Comment by staebler** [comment]
The operator will need permission to read the cloud config ConfigMap from the openshift-config-managed namespace. See https://github.com/openshift/cluster-storage-operator/pull/112.

File: assets/controller.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml:57]: looks better, thanks!

[the remark this answers is not in the document]
Comment (jsafrane) [reply assets/controller.yaml:57]: Wait, it's still `/ca`.

**Comment by jsafrane** [comment]
CI failed:
```
level=info msg=Cluster operator storage Available is False with AWSEBSCSIDriverOperatorCR_WaitForOperator::AWSEBSCSIDriverOperatorDeployment_WaitDeployment: AWSEBSCSIDriverOperatorCRAvailable: Waiting for AWSEBS operator to report status
level=info msg=AWSEBSCSIDriverOperatorDeploymentAvailable: Waiting for a Deployment pod to start
```

And the operator logs:
```
F1204 17:52:23.400935       1 starter.go:140] could not determine if a custom CA bundle is in use: failed to get the openshift-config-managed/kube-cloud-config ConfigMap: configmaps "kube-cloud-config" is forbidden: User "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-operator" cannot get resource "configmaps" in API group "" in the namespace "openshift-config-managed"
goroutine 253 [running]:
k8s.io/klog/v2.stacks(0xc000134001, 0xc00003c000, 0x18c, 0x1dd)
	k8s.io/klog/v2@v2.3.0/klog.go:996 +0xb9
k8s.io/klog/v2.(*loggingT).output(0x337b480, 0xc000000003, 0x0, 0x0, 0xc000c2e230, 0x32a87db, 0xa, 0x8c, 0x0)
	k8s.io/klog/v2@v2.3.0/klog.go:945 +0x191
k8s.io/klog/v2.(*loggingT).printf(0x337b480, 0xc000000003, 0x0, 0x0, 0x220c1bf, 0x37, 0xc000a15d40, 0x1, 0x1)
	k8s.io/klog/v2@v2.3.0/klog.go:733 +0x17a
k8s.io/klog/v2.Fatalf(...)
	k8s.io/klog/v2@v2.3.0/klog.go:1463
github.com/openshift/aws-ebs-csi-driver-operator/pkg/operator.withCustomCABundle(0x2314438, 0x252e140, 0xc0009351e0, 0x252e140)
	github.com/openshift/aws-ebs-csi-driver-operator/pkg/operator/starter.go:140 +0xd3
github.com/openshift/aws-ebs-csi-driver-operator/pkg/operator.RunOperator(0x24fe6c0, 0xc000b52580, 0xc0009f67b0, 0x0, 0x0)
	github.com/openshift/aws-ebs-csi-driver-operator/pkg/operator/starter.go:88 +0x6c8
github.com/openshift/library-go/pkg/controller/controllercmd.ControllerBuilder.getOnStartedLeadingFunc.func1.1(0xc0003d9a40, 0xc000142f00, 0x24fe6c0, 0xc000b52580, 0xc0009f67b0)
	github.com/openshift/library-go@v0.0.0-20201006115306-ed636feadb9c/pkg/controller/controllercmd/builder.go:309 +0x83
created by github.com/openshift/library-go/pkg/controller/controllercmd.ControllerBuilder.getOnStartedLeadingFunc.func1
	github.com/openshift/library-go@v0.0.0-20201006115306-ed636feadb9c/pkg/controller/controllercmd/builder.go:307 +0xb6
```

The operator IMO needs new RBAC. Which is going to be funny, because CSO must be updated first.

**Comment by jsafrane** [comment]
(and it should not crash on error...)

File: pkg/operator/starter.go [hunk 2]
```diff
@@ -91,15 +102,97 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
...
+}
+
+// withCustomCABundle executes the asset as a template to fill out the parts required when using a custom CA bundle.
+// The `caBundleConfigMap` parameter specifies the name of the ConfigMap containing the custom CA bundle. If the
+// argument supplied is empty, then no custom CA bundle will be used.
+func withCustomCABundle(assetFunc func(string) []byte, kubeClient kubeclient.Interface) func(string) []byte {
+	templateData := controllerTemplateData{}
+	switch used, err := isCustomCABundleUsed(kubeClient); {
+	case err != nil:
+		klog.Fatalf("could not determine if a custom CA bundle is in use: %v", err)
```
Comment (jsafrane) [inline pkg/operator/starter.go:140]: This will panic on random API server errors, which is bad.
@bertinatto do you have idea how to work around this?

**Comment by staebler** [comment]
> (and it should not crash on error...)

Why do you think that it should not crash on error? Crashlooping the pod when it cannot initialize does not seem unreasonable.

**Comment by staebler** [comment]
> The operator IMO needs new RBAC. Which is going to be funny, because CSO must be updated first.

Yes, there is a [PR](openshift/cluster-storage-operator#112) for that.

File: assets/controller.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (staebler) [reply assets/controller.yaml]: Oops. I guess I never pushed that change. Done now.

[the remark this answers is not in the document]
Comment (bertinatto) [reply assets/controller.yaml]: hm... these asset files aren't really templates, we've been using string replacers instead.

File: pkg/operator/starter.go (hunk 2 above)
Comment (bertinatto) [reply pkg/operator/starter.go:140]: Ideally we shouldn't do any replacement during the start up. That's better done in a controller

File: pkg/operator/starter.go [hunk 3]
```diff
@@ -91,15 +102,97 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
...
+}
+
+// isCustomCABundleUsed returns true if the cloud config ConfigMap exists and contains a custom CA bundle.
+func isCustomCABundleUsed(kubeClient kubeclient.Interface) (bool, error) {
+	cloudConfigCM, err := kubeClient.CoreV1().
+		ConfigMaps(cloudConfigNamespace).
+		Get(context.Background(), cloudConfigName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		// no cloud config ConfigMap so there is no CA bundle
+		return false, nil
```
Comment (bertinatto) [inline pkg/operator/starter.go:191]: I think assuming the api server will be available during the operator startup can be problematic

**Comment by staebler** [comment]
9c39875...7bdcf9c
* added openshift-config-managed to the list of namespaces that the kube informers are watching

**Pull request merged**